opencog/timeoctomap/pointmemory/PointMemorySCM.cc

```diff
@@ -0,0 +1,360 @@
+  long d_mil=chrono::duration_cast<chrono::milliseconds>(tp.time_since_epoch()).count();
+  long mil_diff=d_mil-d_sec*1000;
+  ts+=to_string(mil_diff);
+  return Handle(createLink(AT_TIME_LINK,Handle(createNode(TIME_NODE,ts)),ato));
+}
+
+Handle PointMemorySCM::get_last_ato(const string& map_name,Handle ato,int elapse)
+{
+  time_pt tpt,tp; duration_c dr;
+  tsa[map_name]->get_current_time_range(tpt,dr);
+  dr=std::chrono::milliseconds(elapse);
+  tpt-=dr;
+  bool r=tsa[map_name]->get_last_time_elapse_atom_observed(ato,tpt,tp);
+  if (!r) return UndefinedHandle;
+  //make and return atTimeLink
+  std::time_t ttp=std::chrono::system_clock::to_time_t(tp);
+  //string ts = std::put_time(std::localtime(&ttp), "%F %T ");
+  char buff[31];
+  strftime(buff, 30, "%Y-%m-%d %H:%M:%S ", std::localtime(&ttp));
+  string ts(buff);
+
+  //time_since_epoch gives duration, duration to seconds and milliseconds then subtract [millisec - sec to millisec] 
+  //add milli sec to ts
+  long d_sec=chrono::duration_cast<chrono::seconds>(tp.time_since_epoch()).count();
+  long d_mil=chrono::duration_cast<chrono::milliseconds>(tp.time_since_epoch()).count();
+  long mil_diff=d_mil-d_sec*1000;
+  ts+=to_string(mil_diff);
+  return Handle(createLink(AT_TIME_LINK,Handle(createNode(TIME_NODE,ts)),ato));
+}
+
+Handle PointMemorySCM::get_at_loc_ato(const string& map_name,double x,double y,double z)
+{
+  Handle ato;
+  if (tsa[map_name]->get_atom_current_time_at_location(point3d(x,y,z),ato))
+     return ato;
+  return UndefinedHandle;
+}
+
+Handle PointMemorySCM::get_past_loc_ato(const string& map_name,int elapse,
+                            double x,double y,double z)
+{
+  time_pt tpt; duration_c dr;
+  tsa[map_name]->get_current_time_range(tpt,dr);
+  dr=std::chrono::milliseconds(elapse);
+  tpt-=dr;
+  Handle ato;
+  if (tsa[map_name]->get_atom_at_time_by_location(tpt,point3d(x,y,z),ato))
+     return ato;
+  return UndefinedHandle;
+}
+
+Handle PointMemorySCM::get_locs_ato(const string& map_name,Handle ato)//listlink atLocationLink
+{
+  point3d_list pl=tsa[map_name]->get_locations_of_atom_occurence_now(ato);
+  if (pl.size()<1)return UndefinedHandle;
+  HandleSeq loc_links(pl.size());
+  while (pl.size()>0)
+  {
+    point3d pt=pl.front();
+    loc_links.push_back(Handle(createLink(AT_LOCATION_LINK,
+      Handle(createNode(OBJECT_NODE,map_name)),
+      Handle(createLink(LIST_LINK,ato,
+        Handle(createLink(LIST_LINK,
+          Handle(createNode(NUMBER_NODE,to_string(pt.x()))),
+          Handle(createNode(NUMBER_NODE,to_string(pt.y()))),
+          Handle(createNode(NUMBER_NODE,to_string(pt.z())))
+        ))
+      ))
+    )));
+    pl.pop_front();
+  }//while
+  
+  return Handle(createLink(LIST_LINK,loc_links));
+}
+
+Handle PointMemorySCM::get_past_locs_ato(const string& map_name,Handle ato,int elapse)
+{
+  time_pt tpt; duration_c dr;
+  tsa[map_name]->get_current_time_range(tpt,dr);
```

Comment:
@yantrabuddhi Wait, don't tell you indent yourself! Again, curious what editor you use.